Bots follow you when you press +use on them

sunzenshen opened the pull request:
## Description
Simple snake formation while copying thermoptic and crouch actions. If more than one bot is following, they will choose a neighbor to follow to reduce collisions.

## Toolchain
- Windows MSVC VS2022

## Linked Issues

sunzenshen commented:
https://github.com/user-attachments/assets/bea3dff0-5e1a-45ab-ac95-fbd76d3bc560

https://github.com/user-attachments/assets/5579993f-bd89-49ba-8748-36793716aa78

AdamTadeusz commented:
https://tenor.com/view/heavy-time-to-dance-conga-tf2-gif-19530172

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_seek_and_destroy.cpp:
```diff
@@ -81,7 +81,66 @@ ActionResult< CNEOBot >	CNEOBotSeekAndDestroy::Update( CNEOBot *me, float interv
+				}
+
+				// Move to the next leader in the chain
+				pPlayerToMirror = pPlayerToMirror->m_hLeadingPlayer.Get();
```
Along the lines of whether it makes sense for the full chain to mimic cloak and crouch behavior, maybe it makes sense only to check the person you are following immediately. (This also reduces the complexity of searching for chain cycles.)

This has the effect that bots behind a Support will not cloak, but that actually might make sense if the Support will always be visible.

Rainyan commented:
(I'm temporarily removing the review request since it's a draft, and the pushes will keep pinging the team members otherwise)

sunzenshen commented:
https://github.com/user-attachments/assets/fc1262f3-9269-47f6-938c-d71478fb81b8

- collecting bots for your current squad star by pressing use on them
- if a bot is in a different squad star, they will change their squad star to match your current one
- highlights which bots are in your squad star with color in the top-left squad star menu
- placing a ping sends bots in your currently selected squad star to that waypoint

https://github.com/user-attachments/assets/a331808d-ad88-4464-bd02-3a77e2f41788

- bots in a squad star attempt to cover surrounding angles using a rudimentary vector repulsion calculation where bots are repulsed from looking at close walls and nearby squad star teammates
- (known fixed issue in the video: the bots had a tendency to look randomly up or down, which was fixed by setting the z dimension of their final look vector to 0)
- collect bots waiting at a squad star waypoint by walking near to have them follow you

Not shown in the videos:
- calibrating how close bots follow you by setting a ping near your feet. distance clamped between a cvar minimum following distance and a maximum following distance
- releasing bots to follow default behaviors by pressing use on them while your squad star is the same.

sunzenshen commented:
Looked into improving the bot spacing and angle coverage when they arrive at the ghost objective:
- increased arrival zone to prevent bots from re-navigating on top of ghost
- proximity based player repulsion instead of squad based repulsion before, in order to allow non squad members to adjust fanout next to each other
- recalculate path on commander ping to improve responsiveness of repositioning
- reduced ping cooldown for bot following to make bots more responsive to following

The behavior inside of enclosed spaces with windows is admittedly be bit odd in the latter part of this clip:

https://github.com/user-attachments/assets/f39d9c15-3352-4102-ad4a-a5f1c1451bfc

Spaces that are more open with clear sightlines tend to work a bit better:

https://github.com/user-attachments/assets/ce26e9a7-24c1-4a47-af3f-64b86c97f75c

AdamTadeusz commented:
Also I guess this should be a separate PR, but it would be cool if the squad leader was sneaking, for the bots to sneak as well as long as they're fairly close,

AdamTadeusz commented:
Once the game crashing in deathmatch with bots is fixed, the bots will react to the player's request to follow them and move to the player's pings in deathmatch

<img width="1922" height="1230" alt="image" src="https://github.com/user-attachments/assets/ccf7e180-3f56-4565-a270-6441bf1f8af0" />

AdamTadeusz commented:
also can the cs2 squad hud get the unique colour for bots following the player too 👉 👈

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_behavior.cpp:
```diff
@@ -722,7 +728,8 @@ void CNEOBotMainAction::FireWeaponAtEnemy( CNEOBot *me )
 				// spray for a bit
-				me->PressFireButton( neo_bot_fire_weapon_min_time.GetFloat() );
+				// NEO Jank: Spraying wastes bullets into friends and walls, so just tap fire
+				me->PressFireButton( /*neo_bot_fire_weapon_min_time.GetFloat()*/ );
```
This is the only significant use case as far as I can tell. I ended up removing this convar.

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_seek_and_destroy.cpp:
```diff
@@ -20,13 +20,19 @@ ConVar neo_bot_disable_seek_and_destroy( "neo_bot_disable_seek_and_destroy", "0"
+ConVar sv_neo_bot_cmdr_look_weights_friendly_repulsion("sv_neo_bot_cmdr_look_weights_friendly_repulsion", "2",
 	FCVAR_NONE, "Weight for friendly bot repulsion force", true, 1, true, 9999);
-ConVar sv_neo_bot_cmdr_look_weights_wall_repulsion("sv_neo_bot_cmdr_look_weights_wall_repulsion", "1",
+ConVar sv_neo_bot_cmdr_look_weights_wall_repulsion("sv_neo_bot_cmdr_look_weights_wall_repulsion", "3",
```
In combination with the decreased wall whisker range, wall whisker checks have shorter range, but count more when the whiskers do touch.

AdamTadeusz commented:
we're gonna have to refactor this at some point down the line so following a leader and their commands is a separate behavior

sunzenshen commented on src/game/server/neo/bot/neo_bot.cpp:
```diff
@@ -2470,6 +2475,57 @@ bool CNEOBot::IsFiring() const
+	return false;
+}
+
 NeoClass CNEOBot::ChooseRandomClass() const
```
In case anyone was wondering what the merge conflict was, the above code block conflicted with https://github.com/NeotokyoRebuild/neo/commit/6351094f7d5a46837b6abff0dd67de344afb2177 at the same location of the below ChooseRandomClass implementation.

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_attack.cpp:
```diff
@@ -58,10 +58,10 @@ ActionResult< CNEOBot >	CNEOBotAttack::Update( CNEOBot *me, float interval )
 
 	if ( bHasRangedWeapon
-		&& (me->IsThreatFiringAtMe(threat->GetEntity()) || threat->IsVisibleInFOVNow())
+		&& me->IsThreatFiringAtMe(threat->GetEntity())
```
After checking that my branch still works okay after a series of rebases, I decided that the bots were a bit too timid (they would run for cover when they see an enemy, but in effect this meant that they would cower around corners).

So I tweaked the logic to have bots only take cover if someone they notice is directly shooting at them.  I feel like this is more analogous to how people actually play.

As part of demoing this change, I recorded some footage for review:
- https://streamable.com/93kzvn
- https://streamable.com/s9yfuy

sunzenshen commented on src/game/client/neo/ui/neo_hud_round_state.cpp:
```diff
@@ -671,6 +667,12 @@ void CNEOHud_RoundState::DrawPlayerList()
 			}
 
+			// Don't include local player or enemies in squad/friendly lists
+			if ((i == localPlayerIndex) || !ArePlayersOnSameTeam(i, localPlayerIndex))
```
I accidentally forgot to count the player as part of the  # vs #" live counters when refactoring the squad lists to show who was selected.

sunzenshen commented:
Turning this into a draft so I can create smaller PRs based on smaller component parts of this large PR.

sunzenshen commented:
Updated this PR to allow using the command feature as a NavMesh debugging tool:

https://github.com/user-attachments/assets/06970d19-a1d8-4f31-b2fc-721f967388f4

New ConVar `sv_neo_bot_cmdr_debug_pause_uncommanded 1;` to allow commanding bots, while pausing every other bot that doesn't have a commander.

sunzenshen commented:
Refactored this PR to remove the changes not related to commanding bots, and moved most of the logic into a new behavior class definition, and moved the transition from CNEOSeekAndDestroy into instead the tactical monitor class.  I think this PR is ready for a second look, but it's also of lower priority in my mind to the other bot PRs.

One side effect of this refactor that reviewers may notice is that squad followers will still take cover if they are under fire, but they no longer follow enemies in the Attack behavior.  This was intentional as I noticed it was easier to manually set up ambushes or to micromanage which cover bots should run for when the bots don't ignore your ping waypoints to run after an enemy.

sunzenshen commented on src/game/client/neo/ui/neo_hud_round_state.cpp:
```diff
@@ -615,6 +615,13 @@ void CNEOHud_RoundState::DrawNeoHudElement()
+	ConVarRef cl_neo_bot_cmdr_enable_ref("sv_neo_bot_cmdr_enable");
+	if (cl_neo_bot_cmdr_enable_ref.IsValid() && cl_neo_bot_cmdr_enable_ref.GetBool())
+	{
+		DrawPlayerList_BotCmdr();
```
Decided that the best way to ensure that there are no regressions possible for the spring tournament, while allowing the use of this feature for use in navmesh debugging, was to hide the related changes behind a feature flag.  The idea is that if sv_neo_bot_cmdr_enable is not enabled, the behavior should be indistinguishable from before this feature is introduced.

sunzenshen replied:
The relevant console commands for testing are:

```
sv_neo_bot_cmdr_enable 1;
sv_neo_bot_cmdr_debug_pause_uncommanded 1;
```

Rainyan replied:
We should probably either return guard the `!cl_neo_bot_cmdr_enable_ref.IsValid()` path too, or simply Assert .IsValid() for debug rather than fail for it silently

Rainyan commented on src/game/client/neo/ui/neo_hud_round_state.cpp:
```diff
@@ -712,8 +719,124 @@ void CNEOHud_RoundState::DrawPlayerList()
+		const int localStar = g_PR->GetStar(localPlayerIndex);
+
+		// Single pass to collect and categorize players
+		for (int i = 0; i < (MAX_PLAYERS + 1); i++)
```
Player entity indices are one-indexed, so we could loop `i = 1; i <= MAX_PLAYERS` to skip the redundant i=0 case

Rainyan commented on src/game/server/neo/bot/behavior/neo_bot_command_follow.cpp:
```diff
@@ -0,0 +1,374 @@
+	}
+
+	if (pCommander->IsAlive())
+	{
```
Nit: this condition goes 4 levels deep (if -> else if -> if -> if ...) which can make it a bit confusing to keep track of the scope. I'd personally prefer if we reversed this base case condition to an early return:
```cpp
if (!pCommander->IsAlive())
{
	// Commander died
	return false; // This triggers OnEnd which clears vars
}
```
so that we can collapse the lowest indent away entirely.

Rainyan commented on src/game/server/neo/bot/behavior/neo_bot_command_follow.h:
```diff
@@ -0,0 +1,34 @@
+	virtual ActionResult< CNEOBot >	OnResume( CNEOBot *me, Action< CNEOBot > *interruptingAction );
+	virtual void					OnEnd( CNEOBot *me, Action< CNEOBot > *nextAction );
+
+	virtual const char *GetName( void ) const	{ return "CommandFollow"; };
```
Nit: would prefer if we explicitly declared these `override`

Rainyan commented on src/game/shared/neo/neo_player_shared.cpp:
```diff
@@ -157,9 +161,43 @@ void CheckPingButton(CNEO_Player* player)
+		if (distSqrToPing < sv_neo_bot_cmdr_stop_distance_sq_max.GetFloat())
+		{
+			// If pinging close to self, calibrate follow distance of commanded bots based on distance to ping
+			player->m_flBotDynamicFollowDistanceSq = Clamp(distSqrToPing, 5000.0f, sv_neo_bot_cmdr_stop_distance_sq_max.GetFloat());
```
Nit: maybe it'd be good to get the clamp min value from the cvar directly, so these cannot go out of sync if someone updates the cvar min value code but forgets about this hard-coded 5000.f

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_command_follow.cpp:
```diff
@@ -0,0 +1,374 @@
+			// follow_stop_distance_sq would be confusing if player doesn't know about distance tuning
+			me->m_hLeadingPlayer = pCommander;
+			m_vGoalPos = vec3_origin;
+			pCommander->m_vLastPingByStar.GetForModify(me->GetStar()) = vec3_origin;
```
Added a new constant VECTOR_INVALID_WAYPOINT

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_pause.cpp:
```diff
@@ -0,0 +1,61 @@
+	if (path)
+	{
+		// Stop existing path
+		const_cast<PathFollower *>(path)->Invalidate();
```
I was kind of overthinking this, nulled out the path in a simpler way.

sunzenshen commented on src/game/shared/neo/neo_player_shared.cpp:
```diff
@@ -157,9 +161,43 @@ void CheckPingButton(CNEO_Player* player)
 	}
 }
 
+ConVar sv_neo_bot_cmdr_enable("sv_neo_bot_cmdr_enable", "0",
```
FCVAR_REPLICATED - I need this flag to be shared with the client so that their playerlist UI can switch between the original player/avatar list behavior or the new highlighted commanded bots version.

FCVAR_ARCHIVE - Allows server operators or single players to decide once if they want to try out the feature, but this feature is still off by default so that parity gameplay is not affected for those not opting into using the feature.

Rainyan commented on src/game/server/neo/bot/behavior/neo_bot_command_follow.cpp:
```diff
@@ -0,0 +1,369 @@
+		float wallRepulsionWeight = sv_neo_bot_cmdr_look_weights_wall_repulsion.GetFloat();
+		vFinalForce = (vBotRepulsion * friendlyRepulsionWeight) + (vWallRepulsion * wallRepulsionWeight);
+		vFinalForce.NormalizeInPlace();
+		vFinalForce.z = 0; // avoid tilting awkwardly up or down
```
This may be fine, but since it seems like we're doing the "unit vector times a scale" pattern, would it make more sense to first zero the Z axis, and only then normalize (since zeroing after the normalization means its length no longer equals one, i.e. the magnitude of the final X and Y depend on what Z was before zeroing)? I'm not really sure so this is more of a question than anything.

Rainyan reviewed (approved):
This is an awesome feature and I wish we can reach consensus to default-enable it eventually.

sunzenshen commented:
> This is an awesome feature and I wish we can reach consensus to default-enable it eventually.

My thinking was that we could wait to enable this feature by default until after the spring tournament build cutoff is released.  By enabling the feature by default after the spring tournament, we can avoid any possible unintended issues cropping up during the the competition, and also benefit from some potential increased attention to advertise the feature.
Then after a few months of baking or even just before anyone wants to make updates to neo_hud_round_state, we could have a PR ready to make the bot version of the round state code the primary path that always happens, and then later on we just completely delete the old version of the round state code when we feel confident about the bot-command-aware approach.

sunzenshen commented on src/game/server/neo/bot/behavior/neo_bot_command_follow.cpp:
```diff
@@ -0,0 +1,369 @@
+ConVar sv_neo_bot_cmdr_look_weights_friendly_max_dist_sq("sv_neo_bot_cmdr_look_weights_friendly_max_dist_sq", "5000",
+	FCVAR_CHEAT, "Distance to compare friendly repulsion forces", true, 1, true, 100000);
+ConVar sv_neo_bot_cmdr_look_weights_wall_repulsion_whisker_dist("sv_neo_bot_cmdr_look_weights_wall_repulsion_whisker_dist", "500",
+	FCVAR_CHEAT, "Distance to extend whiskers", true, 1, true, 100000);
```
In retrospect, I think I must have been embarassingly conflating _sq and flDistSqr, which as you point out the latter was already being handled with a FastSqrt.  I think removing the _sq is the more intuitive choice, as if we went with using MAX_TRACE_LENGTH ^ 2 as the max range, then the default values and configurable range would involve very large numbers, which could be much harder for console users to intuit based on Hammer units.

Rainyan commented on src/game/client/neo/ui/neo_hud_round_state.cpp:
```diff
@@ -712,8 +720,127 @@ void CNEOHud_RoundState::DrawPlayerList()
+		{
+			continue;
+		}
+		if (i == localPlayerIndex || localPlayerSpec || hideDueToScoreboard || !ArePlayersOnSameTeam(i, localPlayerIndex))
```
Nit: It seems like `!ArePlayersOnSameTeam` is redundant, since we already checked `if (playerTeam != localPlayerTeam)` on line 777 above. Maybe it should be removed, since it's going to invoke the virtual call `g_Teams[i]->ContainsPlayer` twice per each iteration, so worst case it's `gpGlobals->maxClients * 2` extra virtual method calls per each frame this draw call happens.

Or if the `ArePlayersOnSameTeam` case is a concern, perhaps we should assert on it for debug builds, rather than silently guard against it with the continue.

sunzenshen replied:
I'll address this in a separate PR.
